metabase/metabase
# Systematize date range formats in Filter name

**Problem:** Accumulated times/time ranges are not properly displayed in the resulting filter

**Context:** 
- Fixes accumulated time unit display formats in: https://github.com/metabase/metabase/issues/12496
- This completes the work started in https://github.com/metabase/metabase/pull/32066 to fix the issue above and to systemically handle all date types.
  - See [<code>DATE_RANGE_FORMAT_SPECS</code>](https://github.com/metabase/metabase/blob/015cfed47ccf03a7013695928de858a041c93369/frontend/src/metabase/lib/formatting/date.tsx#L62-L75) in date.tsx, or the generated table summary below.

---
**Desired Behavior**/**Acceptance Criteria**

_Filter Pill Time Formats_

 <table>
    <td><strong>date unit</strong></td>
    <td><strong>range type</strong></td>
    <td><strong>example</strong></td>
    <tr>
    <td valign=top rowspan="2"><code>year</code></td>
    <td><em>single</em></td>
    <td>2018</td>
    </tr>
    <tr>
    <td><em>otherwise</em></td>
    <td>2018–2019</td>
    </tr>
    <tr>
    <td valign=top rowspan="3"><code>quarter</code></td>
    <td><em>single</em></td>
    <td>Q2 2018</td>
    </tr>
    <tr>
    <td><em>same year</em></td>
    <td>Q2–Q4 2018</td>
    </tr>
    <tr>
    <td><em>otherwise</em></td>
    <td>Q2 2018 – Q3 2019</td>
    </tr>
    <tr>
    <td valign=top rowspan="2"><code>quarter-of-year</code></td>
    <td><em>single</em></td>
    <td>Q2</td>
    </tr>
    <tr>
    <td><em>otherwise</em></td>
    <td>Q2–Q4</td>
    </tr>
    <tr>
    <td valign=top rowspan="3"><code>month</code></td>
    <td><em>single</em></td>
    <td>September 2018</td>
    </tr>
    <tr>
    <td><em>same year</em></td>
    <td>September–December 2018</td>
    </tr>
    <tr>
    <td><em>otherwise</em></td>
    <td>September 2018 – January 2019</td>
    </tr>
    <tr>
    <td valign=top rowspan="2"><code>month-of-year</code></td>
    <td><em>single</em></td>
    <td>September</td>
    </tr>
    <tr>
    <td><em>otherwise</em></td>
    <td>September–December</td>
    </tr>
    <tr>
    <td valign=top rowspan="3"><code>week</code></td>
    <td><em>same month</em></td>
    <td>January 1–21, 2017</td>
    </tr>
    <tr>
    <td><em>same year</em></td>
    <td>January 1 – May 20, 2017</td>
    </tr>
    <tr>
    <td><em>otherwise</em></td>
    <td>January 1, 2017 – February 10, 2018</td>
    </tr>
    <tr>
    <td valign=top rowspan="2"><code>week-of-year</code></td>
    <td><em>single</em></td>
    <td>20th week of the year</td>
    </tr>
    <tr>
    <td><em>otherwise</em></td>
    <td>34th–40th weeks of the year</td>
    </tr>
    <tr>
    <td valign=top rowspan="4"><code>day</code></td>
    <td><em>single</em></td>
    <td>January 1, 2018</td>
    </tr>
    <tr>
    <td><em>same month</em></td>
    <td>January 1–2, 2018</td>
    </tr>
    <tr>
    <td><em>same year</em></td>
    <td>January 1 – February 2, 2018</td>
    </tr>
    <tr>
    <td><em>otherwise</em></td>
    <td>January 1, 2018 – February 2, 2019</td>
    </tr>
    <tr>
    <td valign=top rowspan="2"><code>day-of-week</code></td>
    <td><em>single</em></td>
    <td>Monday</td>
    </tr>
    <tr>
    <td><em>otherwise</em></td>
    <td>Monday – Thursday</td>
    </tr>
    <tr>
    <td valign=top rowspan="2"><code>day-of-month</code></td>
    <td><em>single</em></td>
    <td>20th day of the month</td>
    </tr>
    <tr>
    <td><em>otherwise</em></td>
    <td>10th–12th days of the month</td>
    </tr>
    <tr>
    <td valign=top rowspan="2"><code>day-of-year</code></td>
    <td><em>single</em></td>
    <td>123rd day of the year</td>
    </tr>
    <tr>
    <td><em>otherwise</em></td>
    <td>100th–123rd days of the year</td>
    </tr>
    <tr>
    <td valign=top rowspan="4"><code>hour</code></td>
    <td><em>single</em></td>
    <td>January 1, 2018, 11:00–59 AM</td>
    </tr>
    <tr>
    <td><em>same day</em></td>
    <td>January 1, 2018, 11:00 AM – 2:59 PM</td>
    </tr>
    <tr>
    <td><em>same year</em></td>
    <td>January 1, 11:00 AM – February 2, 2018, 2:59 PM</td>
    </tr>
    <tr>
    <td><em>otherwise</em></td>
    <td>January 1, 2018, 11:00 AM – February 2, 2019, 2:59 PM</td>
    </tr>
    <tr>
    <td valign=top rowspan="2"><code>hour-of-day</code></td>
    <td><em>single</em></td>
    <td>11:00–59 AM</td>
    </tr>
    <tr>
    <td><em>otherwise</em></td>
    <td>11:00 AM – 4:59 PM</td>
    </tr>
    <tr>
    <td valign=top rowspan="4"><code>minute</code></td>
    <td><em>single</em></td>
    <td>January 1, 2018, 11:20 AM</td>
    </tr>
    <tr>
    <td><em>same day</em></td>
    <td>January 1, 2018, 11:20 AM – 2:35 PM</td>
    </tr>
    <tr>
    <td><em>same year</em></td>
    <td>January 1, 11:20 AM – February 2, 2018, 2:35 PM</td>
    </tr>
    <tr>
    <td><em>otherwise</em></td>
    <td>January 1, 2018, 11:20 AM – January 2, 2019, 2:35 PM</td>
    </tr>
    <tr>
    <td valign=top rowspan="2"><code>minute-of-hour</code></td>
    <td><em>single</em></td>
    <td>minute :05</td>
    </tr>
    <tr>
    <td><em>otherwise</em></td>
    <td>minutes :05–30</td>
    </tr>
    </table>

- [ ] The table above should be how all times are displayed in the filter pill
  - e.g. `Day -> Same Month` <img valign=middle width="239" alt="CleanShot 2023-07-04 at 05 54 19@2x" src="https://github.com/metabase/metabase/assets/116838/d85181c9-c05e-4d3e-b108-6a2b17561224">
- [ ] Tests have been added to cover all relevant cases

---
**How to verify**

Try various forms of date filters with various kinds of date types:

1. New question > Raw Data > Sample Database > Accounts
4. Summarize: Count of Rows, by Created At: Day > Visualize
5. Drill into a point or window on the graph.
6. Observe if the created filter label matches the expectations in the table.
7. Repeat with different units. i.e. Created At: Month, Quarter, etc.

---
**Implementation notes**

1. The previous `betterDateLabel()` is replaced with a simpler call to a more featureful date range formatter.
2. The date range formatter now performs a simple lookup in a precomputed format spec object.
3. For a given date range and unit, the date range formatter will search down a list of specs belonging to that unit, then will match the first spec satisfying [moment.isSame](https://momentjs.com/docs/#/query/is-same/) for the two dates in the range, using the specified `same` unit.
4. Embedded in each spec is a test case example.  This is for readability, but it’s also executed as a unit test, since I didn’t want to manage a separate version of the lookup table in the test file.
